### PR TITLE
fix(qwen3-vl): bound CUDA graph caches

### LINKED ISSUE #122: Add CUDA Graph cache management to Qwen3-VL frontends
## Context

Qwen3-VL decode uses CUDA Graph replay, but autoregressive decode has one
annoying property: every generated token advances the decode position. That
changes several runtime values:

- `cache_pos`
- RoPE position
- KV-cache write offset
- attention KV length

The simple and stable way to handle this is to cache one CUDA Graph per decode
position, or per position-like bucket. The current Qwen3-VL frontends follow
that pattern. For example, the new BF16 path uses:

```python
_prefill_graphs[(patch_count, seq_len, span0, span1)]
_decode_graphs[(cache_pos, rope_pos)]
```

This is fine for quickstarts and fixed-shape benchmarks, but it is not bounded.
In a long-running process, many prompt shapes or decode positions can keep
adding graph entries until the process exits.

FlashRT already solves this in some other autoregressive frontends. Qwen36 has
graph-cache helpers, LRU-style OrderedDict storage, clear_graphs(), and
tests such as tests/test_qwen36_nvfp4_graph_cache_bound.py. NexN2 also has a
decode graph cap through FLASHRT_NEXN2_GRAPH_CACHE_MAX.

Qwen3-VL has not adopted that kind of graph-cache management yet.

## Current Qwen3-VL behavior

- Prefill graphs are cached by prompt/image shape.
- Decode graphs are cached by position bucket, such as (cache_pos, rope_pos).
- Repeated buckets replay efficiently.
- New prompt lengths, image shapes, or longer decode ranges capture more graphs.
- Cached graphs are process-local and stay resident until process exit.
- There is no Qwen3-VL-specific cache limit, LRU eviction, clear_graphs() /
  clear_graph_cache() API, or graph-cache stats.

## Why this matters

For fixed local benchmarks, this is acceptable. For serving or arbitrary user
prompts, graph memory can grow over time:

- different prompt lengths can create many prefill graph buckets;
- longer generations can create many decode graph buckets;
- graph memory remains resident for the lifetime of the process;
- users have no way to bound or reset Qwen3-VL graph memory.

This is not only about the new BF16 Orin path. The existing Qwen3-VL RTX/SM89
frontends also use process-local graph dictionaries, so the fix should ideally
be shared across Qwen3-VL frontends.

## Possible follow-ups

- Reuse or factor out the Qwen36-style graph-cache helper pattern for Qwen3-VL.
- Add constructor knobs such as:
    - max_prefill_graphs
    - max_decode_graphs

- Add LRU eviction for prefill/decode graph buckets.
- Add a public clear_graphs() or clear_graph_cache() API.
- Expose graph-cache stats for debugging:
    - number of prefill graph buckets
    - number of decode graph buckets
    - current bucket keys
    - approximate memory use if available

- Add optional warmup for common decode ranges.
- Longer term, investigate position-agnostic decode graphs or lower-level
  cudaGraphExecUpdate, if compatible with the attention backend and PyTorch
  integration.

## Scope

This is only about Qwen3-VL runtime graph-cache management. It should not
change model numerics or the BF16/FP8/NVFP4/INT8 kernel paths.

### PR DESCRIPTION
**Pull request by heiheiha798**
## Summary

Fixes #122.

This PR adds bounded CUDA Graph cache management for all current Qwen3-VL frontends in the repo:

- `Qwen3VlTorchFrontendRtx` (SM120/NVFP4 multimodal path)
- `Qwen3VlTorchFrontendRtxBF16` (Orin SM87 BF16 multimodal path)
- `Qwen3VlFp8Sm89TextFrontend` (SM89 FP8 text path)
- `Qwen3VlFp8Sm89Frontend` (SM89 FP8 multimodal wrapper)

The change replaces unbounded process-local graph dictionaries with LRU-bounded `OrderedDict` caches, adds public `clear_graphs()` APIs, and exposes lightweight `graph_cache_stats()` bucket diagnostics. Prefill graph eviction also drops the corresponding staged/static input buffers so graph entries and buffer residency stay in sync.

Decode graph caches that bake a prompt-dependent RoPE position now key by `(cache_pos, rope_pos)`, avoiding reuse of a graph captured for a different MRoPE continuation position.

The default cap is 256 entries per cache. It can be overridden with:

- `FLASHRT_QWEN3_VL_PREFILL_GRAPH_CACHE_MAX`
- `FLASHRT_QWEN3_VL_DECODE_GRAPH_CACHE_MAX`

Values <= 0 preserve legacy unbounded behavior for callers that intentionally want it.

## Scope

Runtime graph-cache bookkeeping only. This does not change model numerics, kernels, dtype policy, dispatch policy, graph-capture bodies, launch parameters, CMake, or pybind bindings.

## Validation

Environment:

- Python test environment with FlashRT dependencies installed
- Python tests only; no real checkpoint required for the graph-cache regression coverage

Commands:

```bash
PYTHONPATH=. PYTEST_DISABLE_PLUGIN_AUTOLOAD=1 pytest -q \
  tests/test_qwen3_vl_graph_cache.py \
  tests/test_qwen3_vl_rtx_bf16.py \
  tests/test_qwen3_vl_fp8_sm89.py \
  tests/test_qwen3_vl_fp8_sm89_2b.py \
  tests/test_qwen3_vl_smoke.py
python -m compileall -q \
  flash_rt/frontends/torch/qwen3_vl_rtx.py \
  flash_rt/frontends/torch/qwen3_vl_rtx_bf16.py \
  flash_rt/frontends/torch/qwen3_vl_fp8_sm89.py \
  flash_rt/frontends/torch/qwen3_vl_fp8_sm89_multimodal.py \
  tests/test_qwen3_vl_graph_cache.py \
  tests/test_qwen3_vl_rtx_bf16.py \
  tests/test_qwen3_vl_fp8_sm89.py \
  tests/test_qwen3_vl_fp8_sm89_2b.py \
  tests/test_qwen3_vl_smoke.py
git diff --check
rg -n "_decode_graphs\\s*:\\s*dict|_prefill_graphs\\s*:\\s*dict|_pg_buffers\\s*:\\s*dict|_decode_graphs\\s*=\\s*\\{\\}|_prefill_graphs\\s*=\\s*\\{\\}|_pg_buffers\\s*=\\s*\\{\\}" flash_rt/frontends/torch/qwen3_vl*.py tests/test_qwen3_vl_graph_cache.py
```

Results:

- Qwen3-VL Python tests above: 31 passed, 1 skipped
- `compileall`: passed
- `git diff --check`: passed
- Plain unbounded Qwen3-VL graph-cache dict declaration scan: no matches

## Not Run

- Real-checkpoint SM87 BF16 Qwen3-VL smoke.
- Real-checkpoint SM89 Qwen3-VL smoke.
- Real-checkpoint SM120/NVFP4 Qwen3-VL smoke.

Those are useful integration checks for CUDA Graph capture/replay on hardware, but the regression tests in this PR intentionally isolate the cache-management bug without requiring checkpoints or real CUDA Graph capture.

### CODE REVIEW

File: tests/test_qwen3_vl_graph_cache.py [hunk 1]
```diff
@@ -0,0 +1,338 @@
...
+
+    fake_fvk = types.SimpleNamespace(
+        bf16_matmul_bf16=lambda *args: None,
+        bf16_matmul_qwen36_bf16=lambda *args: None)
+    import flash_rt
+    monkeypatch.setitem(sys.modules, "flash_rt.flash_rt_kernels", fake_fvk)
+    monkeypatch.setattr(flash_rt, "flash_rt_kernels", fake_fvk, raising=False)
+    import torch
+    monkeypatch.setattr(torch.cuda, "current_stream",
+                        lambda: _FakeStream(), raising=False)
```
[the remark this answers is not in the document]
Comment (heiheiha798) [reply tests/test_qwen3_vl_graph_cache.py:158]: Fixed in the latest push by monkeypatching `torch.cuda.synchronize` to a no-op in this CI-friendly test, matching the other mocked CUDA surface.

**Review by LiangSu8899 — APPROVED** [review]
